src/biz/casbin/access_control.rs

```diff
@@ -0,0 +1,1009 @@
+  }
+}
+
+#[derive(Clone)]
+pub struct CasbinWorkspaceAccessControl {
+  casbin_access_control: CasbinAccessControl,
+}
+
+impl CasbinWorkspaceAccessControl {
+  pub async fn update_member(&self, uid: &i64, workspace_id: &Uuid, role: AFRole) {
+    let _ = self
+      .casbin_access_control
+      .update(
+        uid,
+        &ObjectType::Workspace(&workspace_id.to_string()),
+        &ActionType::Role(role),
+      )
+      .await;
+  }
+  pub async fn remove_member(&self, uid: &i64, workspace_id: &Uuid) {
+    let _ = self
+      .casbin_access_control
+      .remove(uid, &ObjectType::Workspace(&workspace_id.to_string()))
+      .await;
+  }
+}
+
+#[async_trait]
+impl WorkspaceAccessControl for CasbinWorkspaceAccessControl {
+  async fn get_role_from_uuid(
+    &self,
+    user_uuid: &Uuid,
+    workspace_id: &Uuid,
+  ) -> Result<AFRole, AppError> {
+    let uid = self
+      .casbin_access_control
+      .get_uid(&CollabUserId::UserUuid(user_uuid))
+      .await?;
+    self.get_role_from_uid(&uid, workspace_id).await
+  }
+
+  async fn get_role_from_uid(&self, uid: &i64, workspace_id: &Uuid) -> Result<AFRole, AppError> {
+    let enforcer = self.casbin_access_control.enforcer.read().await;
+    let workspace_id = workspace_id.to_string();
+    let policies = enforcer.get_filtered_policy(
+      POLICY_FIELD_INDEX_OBJECT,
+      vec![ObjectType::Workspace(&workspace_id).to_string()],
+    );
+    // There should only be one entry per user per object, which is enforced in [CasbinAccessControl], so just take one using next.
+    let role = policies
+      .into_iter()
+      .find(|p| p[POLICY_FIELD_INDEX_USER] == uid.to_string())
+      .map(|p| p[POLICY_FIELD_INDEX_ACTION].clone())
+      .and_then(|s| i32::from_str(s.as_str()).ok())
+      .map(AFRole::from)
+      .ok_or(AppError::NotEnoughPermissions(format!(
+        "user:{} is not a member of workspace:{}",
+        uid, workspace_id
+      )));
+
+    role
+  }
+}
+
+#[cfg(test)]
+mod tests {
+  use anyhow::Context;
+  use casbin::{DefaultModel, Enforcer};
+  use shared_entity::dto::workspace_dto::{CreateWorkspaceMember, WorkspaceMemberChangeset};
+
+  use crate::biz::{
+    self,
+    casbin::{adapter::PgAdapter, tests},
+    pg_listener::PgListeners,
+  };
+
+  use super::*;
+
+  #[sqlx::test(migrations = false)]
+  async fn test_workspace_access_control_get_role(pool: PgPool) -> anyhow::Result<()> {
+    tests::setup_db(&pool).await?;
+
+    let model = DefaultModel::from_str(tests::MODEL_CONF).await?;
+    let enforcer = Enforcer::new(model, PgAdapter::new(pool.clone())).await?;
+    let listeners = PgListeners::new(&pool).await?;
+    let access_control = CasbinAccessControl::new(
+      pool.clone(),
+      listeners.subscribe_collab_member_change(),
+      listeners.subscribe_workspace_member_change(),
+      enforcer,
+    );
+    let access_control = access_control.new_workspace_access_control();
+
+    let user = tests::create_user(&pool).await?;
```

Comment:
You can find the handy method to create a user in `tests` folder. Try search the: `edit_workspace_without_permission`